diff --git a/sdk/objc/Framework/Classes/Metal/RTCMTLRenderer.mm b/sdk/objc/Framework/Classes/Metal/RTCMTLRenderer.mm
index 5363aaeda2..3b135ee179 100644
--- a/sdk/objc/Framework/Classes/Metal/RTCMTLRenderer.mm
+++ b/sdk/objc/Framework/Classes/Metal/RTCMTLRenderer.mm
@@ -176,7 +176,8 @@ - (BOOL)setupTexturesForFrame:(nonnull RTCVideoFrame *)frame {
   RTCVideoRotation rotation;
   NSValue *rotationOverride = self.rotationOverride;
   if (rotationOverride) {
-#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
+#if defined(__IPHONE_11_0) && defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && \
+    (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
     if (@available(iOS 11, *)) {
       [rotationOverride getValue:&rotation size:sizeof(rotation)];
     } else

diff --git a/BUILD.gn b/BUILD.gn
index 4420843..f29fbdd 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -350,6 +350,17 @@
       "video",
     ]
 
+    # Additional factory functions to be exposed.
+    # Rational: These factories are small enough (89 KiB / 32+ MiB)
+    #           to be unconditionaly included for user convenience.
+    # That begin said:
+    # TODO(yvesg) Consider making all non-core APIs optional, so that users
+    #             can build a customized library tailored to their needs.
+    deps += [
+      "api/audio_codecs:builtin_audio_decoder_factory",
+      "api/audio_codecs:builtin_audio_encoder_factory",
+    ]
+
     if (build_with_mozilla) {
       deps += [
         "api/video:video_frame",
